clients/geth/specular/rollup/services/sequencer/sequencer.go

```diff
@@ -62,37 +62,126 @@ func New(eth services.Backend, proofBackend proof.Backend, cfg *services.Config,
+		}
+		if prevTx == nil {
+			batchTxs = append(batchTxs, tx)
+		}
+	}
+	return batchTxs
+}
+
+// Send batch
+func (s *Sequencer) sendBatch(batcher *Batcher) {
+	blocks, err := batcher.Batch()
+	if err != nil {
+		log.Crit("Failed to batch blocks", "err", err)
+	}
+	batch := rollupTypes.NewTxBatch(blocks, 0) // TODO: add max batch size
+	s.batchCh <- batch
+}
+
+// Add sorted txs to batch and commit txs
+func (s *Sequencer) addSortedTxsToBatch(batcher *Batcher, sortedTxs *types.TransactionsByPriceAndNonce, batchTxs []*types.Transaction, signer types.Signer) []*types.Transaction {
+	if sortedTxs != nil {
+		for {
+			tx := sortedTxs.Peek()
+			if tx == nil {
+				break
+			}
+			batchTxs = s.modifyTxnsInBatch(batchTxs, tx)
+			sortedTxs.Pop()
+		}
+	}
+	if len(batchTxs) == 0 {
+		return batchTxs
+	}
+	err := batcher.CommitTransactions(batchTxs)
+	if err != nil {
+		log.Crit("Failed to commit transactions", "err", err)
+	}
+	return batchTxs
+}
+
 // This goroutine fetches txs from txpool and batches them
-// TODO: fetching txs through sub is unstable
-// TODO: batch txs by timer and directly querying txpool as worker does
 func (s *Sequencer) batchingLoop() {
 	defer s.Wg.Done()
 	defer close(s.batchCh)
 
+	// Ticker
+	var ticker = time.NewTicker(timeInterval)
+	defer ticker.Stop()
+
 	// Watch transactions in TxPool
 	txsCh := make(chan core.NewTxsEvent, 4096)
 	txsSub := s.Eth.TxPool().SubscribeNewTxsEvent(txsCh)
 	defer txsSub.Unsubscribe()
 
+	// Process txns via batcher
 	batcher, err := NewBatcher(s.Config.Coinbase, s.Eth)
 	if err != nil {
 		log.Crit("Failed to start batcher", "err", err)
 	}
 
+	var batchTxs []*types.Transaction
+
+	// Loop over txns
 	for {
 		select {
-		case ev := <-txsCh:
-			// New transactions from TxPool
-			err = batcher.CommitTransactions(ev.Txs)
-			if err != nil {
-				log.Crit("Failed to commit transactions", "err", err)
+		case <-ticker.C:
+			// Get pending txs - locals and remotes, sorted by price
+			var txs []*types.Transaction
+			signer := types.MakeSigner(batcher.chainConfig, batcher.header.Number)
+
+			pending := s.Eth.TxPool().Pending(true)
+			localTxs, remoteTxs := make(map[common.Address]types.Transactions), pending
+			for _, account := range s.Eth.TxPool().Locals() {
+				if txs = remoteTxs[account]; len(txs) > 0 {
+					delete(remoteTxs, account)
+					localTxs[account] = txs
+				}
 			}
-			blocks, err := batcher.Batch()
-			if err != nil {
-				log.Crit("Failed to batch blocks", "err", err)
+			if len(localTxs) > 0 {
+				sortedTxs := types.NewTransactionsByPriceAndNonce(signer, localTxs, batcher.header.BaseFee)
+				batchTxs = s.addSortedTxsToBatch(batcher, sortedTxs, batchTxs, signer)
+			}
+			if len(remoteTxs) > 0 {
+				sortedTxs := types.NewTransactionsByPriceAndNonce(signer, remoteTxs, batcher.header.BaseFee)
+				batchTxs = s.addSortedTxsToBatch(batcher, sortedTxs, batchTxs, signer)
+			}
+			if len(batchTxs) > 0 {
+				s.sendBatch(batcher)
+			}
+			batchTxs = nil
```

Comment:
nit: can we just init the array at the top of each case for garbage collection instead of assigning nil